sway-lib-std/src/inputs.sw

```diff
@@ -113,12 +157,149 @@ pub fn input_predicate_data<T>(index: u64) -> T {
+    let type = input_type(index);
     match type {
-        Transaction::Script => __gtf::<u8>(0, GTF_SCRIPT_INPUTS_COUNT),
-        Transaction::Create => __gtf::<u8>(0, GTF_CREATE_INPUTS_COUNT),
+        Input::Coin => Option::Some(ContractId::from(__gtf::<b256>(index, GTF_INPUT_COIN_ASSET_ID))),
+        Input::Message => Option::Some(BASE_ASSET_ID),
+        Input::Contract => return Option::None,
+    }
+}
+
+/// Get the witness index from the input at `index`.
+/// If the input's type is `InputCoin` or `InputMessage`,
+/// return the index as an Option::Some(u8).
+/// Otherwise, returns Option::None.
+pub fn input_witness_index(index: u64) -> Option<u8> {
+    let type = input_type(index);
+    match type {
+        Input::Coin => Option::Some(__gtf::<u8>(index, GTF_INPUT_COIN_WITNESS_INDEX)),
+        Input::Message => Option::Some(__gtf::<u8>(index, GTF_INPUT_MESSAGE_WITNESS_INDEX)),
+        Input::Contract => Option::None,
+    }
+}
+
+/// Get the predicate length from the input at `index`.
+/// If the input's type is `InputCoin` or `InputMessage`,
+/// return the length as an Option::Some(u16).
+/// Otherwise, returns Option::None.
+pub fn input_predicate_length(index: u64) -> Option<u16> {
+    let type = input_type(index);
+    match type {
+        Input::Coin => Option::Some(__gtf::<u16>(index, GTF_INPUT_COIN_PREDICATE_LENGTH)),
+        Input::Message => Option::Some(__gtf::<u16>(index, GTF_INPUT_MESSAGE_PREDICATE_LENGTH)),
+        Input::Contract => Option::None,
+    }
+}
+
+/// Get the predicate pointer from the input at `index`.
+/// If the input's type is `InputCoin` or `InputMessage`,
+/// return the pointer as an Option::Some(ptr).
+/// Otherwise, returns Option::None.
+pub fn input_predicate_pointer(index: u64) -> Option<raw_ptr> {
+    let type = input_type(index);
+    match type {
+        Input::Coin => Option::Some(__gtf::<raw_ptr>(index, GTF_INPUT_COIN_PREDICATE)),
+        Input::Message => Option::Some(__gtf::<raw_ptr>(index, GTF_INPUT_MESSAGE_PREDICATE)),
+        Input::Contract => Option::None,
+    }
+}
+
+/// Get the predicate from the input at `index`.
+/// If the input's type is `InputCoin` or `InputMessage`,
+/// return the data, otherwise reverts.
+pub fn input_predicate(index: u64) -> Bytes {
+    let data = input_predicate_pointer(index);
+    let length = input_predicate_length(index).unwrap();
+    let mut data_bytes = Bytes::with_capacity(length);
+
+    match data {
+        Option::Some(d) => {
+            data_bytes.len = length;
+            d.copy_bytes_to(data_bytes.buf.ptr, length);
+            data_bytes
+        },
+        Option::None => revert(0),
+    }
+}
+
+/// Get the predicate data length from the input at `index`.
+/// If the input's type is `InputCoin` or `InputMessage`,
+/// return the data length as an Option::Some(u16).
+/// Otherwise, returns Option::None.
+pub fn input_predicate_data_length(index: u64) -> Option<u16> {
+    let type = input_type(index);
+    match type {
```

Comment:
Remind me, does writing `match input_type(index)` directly not work here?

Reply:
It does indeed. fixed.